src/audio/copier/copier_generic.c

```diff
@@ -107,6 +107,188 @@ int copier_gain_set_fade_params(struct comp_dev *dev, struct dai_data *dd,
 	return 0;
 }
 
+int copier_gain_input16(struct comp_buffer *buff, enum copier_gain_state state,
+			enum copier_gain_envelope_dir dir,
+			struct copier_gain_params *gain_params, uint32_t frames)
+{
+	int16_t *dst = audio_stream_get_rptr(&buff->stream);
+	const int nch = audio_stream_get_channels(&buff->stream);
+	int samples = frames * nch;
+	int16_t gain_env[MAX_GAIN_COEFFS_CNT] = {0};
+	int16_t gain_env_sq;
+	int16_t gain_env_i16;
+	int16_t *dst_tmp;
+	int16_t gain;
+	int nmax, i, j;
+
+	switch (state) {
+	case STATIC_GAIN:
+		/* static gain */
+		if (gain_params->unity_gain)
+			return 0;
+
+		while (samples) {
+			nmax = audio_stream_samples_without_wrap_s16(&buff->stream, dst);
+			nmax = MIN(samples, nmax);
+
+			for (j = 0; j < nch; j++) {
+				dst_tmp = dst + j;
+				gain = gain_params->gain_coeffs[j];
+				for (i = 0; i < nmax; i += nch)
+					dst_tmp[i] = q_multsr_sat_16x16(dst_tmp[i], gain,
+									GAIN_Q10_INT_SHIFT);
+			}
```

Comment:
I'm wondering: with these 2 nested loops, the way it's done here where the external loop is over channels, and it looks like the buffer is organised like "s0ch0, s0ch1, s1ch0, s1ch1,..." we access the buffer with `buf[0], buff[2],... buff[n], buff[1], buff[3],... buff[2*n-1]`. Wouldn't it be more cache-efficient to scan channels inside samples. But then of course we'd have to get a different value of `gain` on every iteration... Probably wouldn't be faster in the end